File: 239.sliding-window-maximum/solution.py
"""
The n depict the number of elements in the nums variable.

TC: O(n log k)
"""
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        results = []
        n = len(nums)
        window = []

        def insert(target: int):
            L = 0
            R = len(window) - 1
            while L <= R:
                mid = (L + R) >> 1
                if window[mid] < target:
                    L = mid + 1
                else:
                    R = mid - 1
            window.insert(L, target)
        
        def remove(target: int):
            L = 0
            R = len(window) - 1
            while L <= R:
                mid = (L + R) >> 1
                if window[mid] < target:
                    L = mid + 1
                else:
                    R = mid - 1
            window.pop(L)


        for i, num in enumerate(nums):
            if i - k >= 0:
                remove(nums[i - k])

            insert(num)

            if i - k + 1 >= 0:
                results.append(window[-1])
                
        return results

File: 239.sliding-window-maximum/solution.ts
function maxSlidingWindow(nums: number[], k: number): number[] {
    let q: number[] = [];
    const results: number[] = [];
    const N = nums.length;
    for (let i = 0; i < N; i++) {
        while ((q.length > 0) && (i - q[0] >= k)) {
            q.shift();
        }
        while ((q.length > 0) && (nums[q.at(-1)!] <= nums[i])) {
            q.pop();
        }
        q.push(i);
        
        if (i < k - 1) {
            continue;
        }

        results.push(nums[q[0]]);
    }
    return results;
};